### DISCUSSION TITLE
Invalid public key

### DISCUSSION THREAD

**Discussion by taaha3244** [langfuse/langfuse#1712] [Support]
Hi everyone,
I have started to integrate Langfuse in my LLM app. I have a setup where i am Loading Docs from Qdrant Database and using LCEL to get response from OPENAI. Here is a brief code for it.

import os
from langchain import hub
from dotenv import load_dotenv
from langchain_community.vectorstores import Qdrant
from qdrant_client import QdrantClient
from qdrant_client.http import models
from qdrant_client import QdrantClient
from langchain_openai import OpenAIEmbeddings
from langchain.prompts import PromptTemplate
from langchain.chains import LLMChain
from langchain_openai import ChatOpenAI
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
from typing import Optional
from langfuse.callback import CallbackHandler
import traceback
# Load environment variables from .env file
load_dotenv()

def util_bot(question: str) -> Optional[str]:
    """
    Processes a given question using a combination of Qdrant vector search and an LLM (Large Language Model) response generation.
    This function embeds the input question using OpenAI's embedding model, searches for relevant context using Qdrant,
    and generates a response based on the context found and the input question using an LLM.

    Args:
        question (str): The user's question to be answered.

    Returns:
        Optional[str]: The generated answer, or None if no answer could be generated.

    Raises:
        Exception: If an error occurs during processing.
    """

    try:
        # Environment variable setup
        openai_api_key = os.getenv('OPENAI_API_KEY')
        qdrant_api_key = os.getenv('QDRANT_API_KEY')
        qdrant_endpoint = os.getenv('QDRANT_CLOUD_URL')
        LANGFUSE_SECRET_KEY=os.getenv('LANGFUSE_API_KEY')
        LANGFUSE_PUBLIC_KEY=os.getenv('LANGFUSE_PUBLIC_KEY')
        LANGFUSE_HOST="https://us.cloud.langfuse.com"

        langfuse_handler = CallbackHandler(secret_key=LANGFUSE_SECRET_KEY,
                                           public_key=LANGFUSE_PUBLIC_KEY,
                                           host=LANGFUSE_HOST)
        
        # Initialization
     
        embeddings_model = OpenAIEmbeddings(model='text-embedding-3-small', openai_api_key=openai_api_key)
        qdrant_client = QdrantClient(url=qdrant_endpoint, api_key=qdrant_api_key)
        qdrant = Qdrant(client=qdrant_client, collection_name="multiple-util-bot-quanntized",
                        embeddings=embeddings_model)
        retriever = qdrant.as_retriever(search_kwargs={"k": 10})

        # Retrieve documents
        docs = retriever.get_relevant_documents(question)

        # Construct the prompt template for the LLM
        prompt = hub.pull("pwoc517/more-crafted-rag-prompt")
        # RAG Chain
        llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0, openai_api_key=openai_api_key)

        def format_docs(docs):
            return "\n\n".join(doc.page_content for doc in docs)

        rag_chain = (
            {"context": retriever | format_docs, "question": RunnablePassthrough()}
            | prompt
            | llm
            | StrOutputParser()
        )

        result = rag_chain.invoke( question, config={"callbacks": [langfuse_handler]})
        return result

    except Exception as e:
        traceback.print_exc()
        return None

But this code is throwing an error:
File "d:\Util-bot-AzureContainerApp\.venv\Lib\site-packages\langfuse\request.py", line 106, in _process_response
    raise APIError(res.status_code, payload)
langfuse.request.APIError: {'error': 'UnauthorizedError', 'message': "Invalid public key. Confirm that you've configured the correct host."} (401): None

I have tried the basic debugging. My env variables are loading correctly. But the error persists with this approach.
However, Inputting the api and public key directly in the call back function runs the code correct.

Thanks in advance for any help

**Comment by marcklingen** [comment]
> However, Inputting the api and public key directly in the call back function runs the code correct.

That's interesting.

Can you print the env when set via dotenv to confirm that it is correct? Langfuse just loads environment variables from os.getenv() if they are not provided in the constructor.

**Reply by taaha3244** [reply]
Yes env variables are being printed correctly
